main.py

```
@@ -604,0 +1,1133 @@
+                    url, attempt, conn_err,
+                    f"Connection error on attempt {attempt+1}/{self.max_retries}: {conn_err}",
+                    f"Connection error persisted for {url} after {self.max_retries} attempts."
+                )
+            except requests.exceptions.Timeout as timeout_err:
+                self._handle_request_error(
+                    url, attempt, timeout_err,
+                    f"Timeout on attempt {attempt+1}/{self.max_retries}: {timeout_err}",
+                    f"Request to {url} timed out after {self.max_retries} attempts."
+                )
+            except Exception as err:
+                logger.error(f"An unexpected error occurred: {err}")
```

Ruff (0.8.2)

Use `logging.exception` instead of `logging.error`

Replace with `exception`

(TRY400)